### PR TITLE
Use 64-bit OOP for systems with 32GiB+ memory

### PR DESCRIPTION
**Pull request by sharwell**

### CODE REVIEW

File: src/VisualStudio/Core/Def/Implementation/Remote/RemoteHostClientServiceFactory.RemoteHostClientService.cs [hunk 1]
```diff
@@ -187,6 +188,12 @@ private void SetRemoteHostBitness()
                         WellKnownExperimentNames.RoslynOOP64bit);
                 }
 
+                if (!x64)
+                {
+                    // Use 64-bit OOP for systems with 32GiB+ memory
```
Comment (genlu) [inline src/VisualStudio/Core/Def/Implementation/Remote/RemoteHostClientServiceFactory.RemoteHostClientService.cs]: Why 32?

File: src/VisualStudio/Core/Def/Implementation/Remote/RemoteHostClientServiceFactory.RemoteHostClientService.cs [hunk 2]
```diff
@@ -187,6 +188,12 @@ private void SetRemoteHostBitness()
...
+                    x64 = MemoryHelper.GetMaxPhysicalMemory() >= 32UL * 1024 * 1024 * 1024;
```
Comment (genlu) [inline src/VisualStudio/Core/Def/Implementation/Remote/RemoteHostClientServiceFactory.RemoteHostClientService.cs]: nit: use shift op?

File: src/VisualStudio/Core/Def/Implementation/Remote/RemoteHostClientServiceFactory.RemoteHostClientService.cs (hunk 1 above)
Comment (sharwell) [reply src/VisualStudio/Core/Def/Implementation/Remote/RemoteHostClientServiceFactory.RemoteHostClientService.cs]: I've seen the 64-bit OOP use over 16GiB while working on Roslyn, but never more than 24.

File: src/VisualStudio/Core/Def/Implementation/Remote/RemoteHostClientServiceFactory.RemoteHostClientService.cs (hunk 2 above)
Comment (sharwell) [reply src/VisualStudio/Core/Def/Implementation/Remote/RemoteHostClientServiceFactory.RemoteHostClientService.cs]: I find this form more readable. Off the top of my head, I'm not sure how many places I need to shift to get 1GiB. 32GiB would be 5 more than that... but still unknown.

**Comment by CyrusNajmabadi** [comment]
Note: i have 32GB of ram, and would like this to apply to me :)  is this >= 32GB, or >32GB? :)

File: src/VisualStudio/Core/Def/Implementation/Remote/RemoteHostClientServiceFactory.RemoteHostClientService.cs (hunk 2 above)
Comment (genlu) [reply src/VisualStudio/Core/Def/Implementation/Remote/RemoteHostClientServiceFactory.RemoteHostClientService.cs]: Hmm, maybe I was confused about the definition of GiB. Isn't one GiB just 2^30? bytes?

Comment (sharwell) [reply src/VisualStudio/Core/Def/Implementation/Remote/RemoteHostClientServiceFactory.RemoteHostClientService.cs]: It is 2<sup>30</sup>, but not everyone knows 30 in the right constant number there.

File: src/VisualStudio/Core/Def/Implementation/Remote/RemoteHostClientServiceFactory.RemoteHostClientService.cs (hunk 1 above)
Comment (heejaechang) [reply src/VisualStudio/Core/Def/Implementation/Remote/RemoteHostClientServiceFactory.RemoteHostClientService.cs]: probably want to use experiment service to do this A/B testing?

File: src/Compilers/Server/VBCSCompiler/MemoryHelper.cs [hunk 3]
```diff
@@ -44,11 +46,20 @@ public static bool IsMemoryAvailable()
...
             }
 
+            ulong max = status.MaxVirtual;
             CompilerServerLogger.Log("Free memory: {1}{0} of {2}{0}.", unit, free >> shift, max >> shift);
+#endif
 
             return free >= 800 << 20; // Value (500MB) is arbitrary; feel free to improve.
         }
 
+        public static ulong GetMaxPhysicalMemory()
```
Comment (jasonmalinowski) [inline src/Compilers/Server/VBCSCompiler/MemoryHelper.cs]: What does this do for VMs with dynamic memory enabled?

File: src/Workspaces/Core/Portable/Microsoft.CodeAnalysis.Workspaces.csproj [hunk 4]
```diff
@@ -29,6 +29,7 @@
   </ItemGroup>
   <ItemGroup Label="Linked Files">
     <Compile Include="..\..\..\CodeStyle\Core\CodeFixes\FixAllContextHelper.cs" Link="CodeFixes\FixAllOccurrences\FixAllContextHelper.cs" />
+    <Compile Include="..\..\..\Compilers\Server\VBCSCompiler\MemoryHelper.cs" Link="Utilities\CompilerUtilities\MemoryHelper.cs" />
```
Comment (jasonmalinowski) [inline src/Workspaces/Core/Portable/Microsoft.CodeAnalysis.Workspaces.csproj]: Why are we linking this into the Workspaces layer? The file is using P/Invokes and therefore is very non-portable. I'd be very worried somebody calling this on a non-Windows platform.

Comment (sharwell) [reply src/Workspaces/Core/Portable/Microsoft.CodeAnalysis.Workspaces.csproj]: I'll move it up to the VS layer

**Pull request closed without merging**